### PR TITLE
[v7r1] Enabling Processes and using concurrent futures

### PR DESCRIPTION
**Pull request by zmathe**
BEGINRELEASENOTES

*Core
CHANGE: Replace ThreadPool to ThreadPoolExecutor(native python)

ENDRELEASENOTES

### CODE REVIEW

File: Core/DISET/ServiceReactor.py [hunk 1]
```diff
@@ -119,24 +122,61 @@ def __createListeners(self):
       self.__listeningConnections[serviceName]['socket'] = transport.getSocket()
     return S_OK()
 
+  def stopChildProcesses(self, _sig, frame):
+    """
+    It is used to properly stop the service when more than one process are used.
+    In principle this is doing the job of runsv, becuase runsv only send a sigterm to the parent process...
+    :param int _sig: the signal sent to the process
```
Comment (andresailer) [inline Core/DISET/ServiceReactor.py]: Need an empty line here

File: Core/DISET/private/GatewayService.py [hunk 2]
```diff
@@ -15,26 +15,35 @@
...
 
-__RCSID__ = "$id:"
 
 import sys
 import cStringIO
+import os
+
+# TODO: Remove ThreadPool later 
+useThreadPoolExecutor = False
+if os.getenv('DIRAC_USE_NEWTHREADPOOL', 'NO').lower() in ('yes', 'true'):
```
Comment (andresailer) [inline Core/DISET/private/GatewayService.py]: Please add an entry to docs/source/AdministratorGuide/ServerInstallations/environment_variable_configuration.rst

File: Core/DISET/private/GatewayService.py [hunk 3]
```diff
@@ -15,26 +15,35 @@
...
+  from concurrent.futures import ThreadPoolExecutor
+  useThreadPoolExecutor = True
+else:
+  from DIRAC.Core.Utilities.ThreadPool import ThreadPool
+  from DIRAC.Core.DISET.private.MessageBroker import MessageBroker, getGlobalMessageBroker
```
Comment (andresailer) [inline Core/DISET/private/GatewayService.py]: Is this import location in the `else` correct? It seems to me that the MessageBroker and getGlobalMessageBroker are use regardless of OldThreadPool or NewThreadPool given lines 96ff in this file. Or maybe lines 96ff need updating.

Comment (zmathe) [reply Core/DISET/private/GatewayService.py]: Ops. You are right. I did a mistake when I cherry-pick (I did not resolved the conflicts correctly) from integration to v7r1:
https://github.com/zmathe/DIRAC/blob/threadPool/Core/DISET/private/GatewayService.py#L25

Comment (zmathe) [reply Core/DISET/private/GatewayService.py]: I will run a test later today again to make sure everything is correct

Comment (andresailer) [reply Core/DISET/private/GatewayService.py]: I was wondering if we can add the integrations tests with DIRAC_USE_NEWTHREADPOOL=yes to the actions

Comment (chrisburr) [reply Core/DISET/private/GatewayService.py]: I think you can do it using something like this: https://github.com/DIRACGrid/DIRAC/commit/1a3a5b3bfe843f57b0f53fe7590e7c10df598255

Comment (zmathe) [reply Core/DISET/private/GatewayService.py]: I need to run with the flag the following test :Integration tests / Integration (5.7, slc6) (pull_request)

Comment (chrisburr) [reply Core/DISET/private/GatewayService.py]: I've done some local testing and an additional change is needed to propagated it to the `client` and `server` containers: d48342cff570509dd0ae2c1174f48e7f20202421

Comment (zmathe) [reply Core/DISET/private/GatewayService.py]: Thanks. I have added your suggestion

File: Core/DISET/private/ServiceConfiguration.py [hunk 4]
```diff
@@ -1,110 +1,115 @@
...
       return 15
 
-  def getCloneProcesses( self ):
+  def getCloneProcesses(self):
     try:
-      return int( self.getOption( "CloneProcesses" ) )
+      return int(self.getOption("CloneProcesses"))
     except:
-      return 1
+      return 0
```
Comment (fstagni) [inline Core/DISET/private/ServiceConfiguration.py]: Why?

Comment (zmathe) [reply Core/DISET/private/ServiceConfiguration.py]: This is the number of how many processes needs to be started. By default we do not want to start any processes

**Comment by atsareg** [comment]
Is this PR ready to go ? I guess that the default behavior is not changed and the new functionality is only used if the env variable is set.

**Comment by fstagni** [comment]
@atsareg  this PR can be merged, please do it today.

**Pull request merged**